packages/rum-core/src/domain/contexts/pageStateContexts.ts

```diff
@@ -0,0 +1,141 @@
+import type { Duration, RelativeTime } from '@datadog/browser-core'
+import {
+  isExperimentalFeatureEnabled,
+  noop,
+  elapsed,
+  addDuration,
+  addEventListeners,
+  relativeNow,
+  DOM_EVENT,
+} from '@datadog/browser-core'
+
+export const PAGE_STATE_TIMELINE_MAX_LENGTH = 500
+
+export const enum PageState {
+  ACTIVE = 'active',
+  PASSIVE = 'passive',
+  HIDDEN = 'hidden',
+  FROZEN = 'frozen',
+  TERMINATED = 'terminated',
+}
+
+type PageStateTimeline = Array<{ state: PageState; startTime: RelativeTime }>
+
+type Counter = { count: number; duration: Duration }
+export type PageStateContext = {
+  active: Counter
+  passive: Counter
+  hidden: Counter
+  frozen: Counter
+  terminated: Counter
+}
+
+export interface PageStateContexts {
+  getPageStates: (startTime: RelativeTime, duration: Duration) => PageStateContext | undefined
+  stop: () => void
+}
+
+let pageStateTimeline: PageStateTimeline = []
+let state: PageState | undefined
+
+export function startPageStateContexts(): PageStateContexts {
+  if (!isExperimentalFeatureEnabled('resource_page_states')) {
+    return {
+      getPageStates: () => undefined,
+      stop: noop,
+    }
+  }
+
+  addPageState(getState())
+
+  const { stop } = addEventListeners(
+    window,
+    [
+      DOM_EVENT.PAGE_SHOW,
+      DOM_EVENT.FOCUS,
+      DOM_EVENT.BLUR,
+      DOM_EVENT.VISIBILITY_CHANGE,
+      DOM_EVENT.RESUME,
+      DOM_EVENT.FREEZE,
+      DOM_EVENT.PAGE_HIDE,
+    ],
+    (event) => {
+      if (!event.isTrusted) return
+
+      if (event.type === DOM_EVENT.FREEZE) addPageState(PageState.FROZEN)
+      else if (event.type === DOM_EVENT.PAGE_HIDE)
+        addPageState((event as PageTransitionEvent).persisted ? PageState.FROZEN : PageState.TERMINATED)
+      else addPageState(getState())
+    },
+    { capture: true }
+  )
+
+  function findStartIndex(startTime: RelativeTime) {
+    let i = pageStateTimeline.length - 1
+    while (i >= 0 && pageStateTimeline[i].startTime > startTime) {
+      i--
+    }
+
+    return i >= 0 ? i : 0
+  }
+
+  return {
+    getPageStates(startTime: RelativeTime, duration: Duration) {
+      const endTime = addDuration(startTime, duration)
+      if (pageStateTimeline.length === 0 || pageStateTimeline[0].startTime >= endTime) return
+
+      const pageStateContext: PageStateContext = {
+        active: { count: 0, duration: 0 as Duration },
+        passive: { count: 0, duration: 0 as Duration },
+        hidden: { count: 0, duration: 0 as Duration },
+        frozen: { count: 0, duration: 0 as Duration },
+        terminated: { count: 0, duration: 0 as Duration },
+      }
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏What is the goal of collecting the count/duration of each states? should not it be more helpful to collect the corresponding timeline directly?

Reply:
The timeline would be an array of object which is hard to query in Datadog.

Reply:
oh, you plan to query that, I would have expect to use this debug data only when looking at a resource with long duration.

Not sure what you would expect to get from those queries though.
Could you share your plans around that?

Reply:
Nothing fancy, my idea was to look at resource with long duration to identify potential pattern and then validate my assumption by queying on one specific state duration for instance to have an idea of volume

Reply:
Get the raw data